### ISSUE TITLE
Add cargo-add (from cargo-edit) to cargo proper

### ISSUE THREAD

**Issue by killercup** [rust-lang/cargo#5586]
Cargo as well as cargo-edit have gone a long way since #4 was opened. It might now make sense to move (parts of) cargo-edit into cargo itself. Basically, we now have a format-preserving TOML library that -- at least for adding a dependency line -- work quite well (you can test this with `cargo install cargo-edit --vers 0.3.0-beta.1 -f`).

Speaking with @matklad, I believe I now have a somewhat solid understanding of what needs to happen to add a new built-in cargo subcommand. One important realization was that cargo-install already contains some of the parts we need.

Here are some steps to get started:

1. Create a `src/bin/cargo/commands/add.rs` (as duplicate of `install.rs` at first)
2. In `src/bin/cargo/commands/mod.rs`, add `add` to `builtin`/`builtin_exec`, and to the list of `mod`s
3. Create a `src/bin/cargo/ops/cargo_add.rs` (as duplicate of `cargo_install.rs` at first) 

One thing we should aim for though, is to now just copy-paste the whole of cargo-edit but to re-use as much of cargo's infrastructure as possible. For example, we should try to use cargo's way of querying the registry as well as its CLI args handling and output formatting. We can, for example, also use `args.workspace.current_manifest` to get the manifest and deal with workspace.

**Comment by killercup** [comment]
@rust-lang/cargo, is this something you'd be interested in being worked on right now? This might work as a mentored even if you can't spend much time to work on it ourselves.

**Comment by alexcrichton** [comment]
I'd personally be thrilled to see this enter Cargo itself! I would ideally prefer to avoid having two TOML parsers built in but if we need to have that for the time being I think it's not the end of the world. (I don't know much about the technical design of cargo-edit right now)

**Comment by steveklabnik** [comment]
This is the number one thing I want in Cargo; a simple 👍 is not enough. Tons and tons and tons of people are very excited for this 🎊

**Comment by ibaryshnikov** [comment]
I'd like to implement it

**Comment by japaric** [comment]
I know that lots of people (myself included) are plenty busy with stuff for the upcoming edition release and that we don't have much spare ~human~ Rustacean resources but I still want to say that it would be AMAZING to have `cargo add` in Cargo 2018. As [mentioned by boats](https://boats.gitlab.io/blog/post/my-experience-with-rust-2018/), `cargo add` fits perfectly with the module / path changes of the edition. And I fully agree, but to me the biggest win would be *not* having to `cargo install cargo-edit`, which involves compiling / linking to C code and can be a [(very) frustrating experience](https://mobile.twitter.com/therealjpster/status/1037646367102050304).

@rust-lang/cargo @killercup @aturon could we make this happen, somehow? :pray:

**Comment by steveklabnik** [comment]
Huuuuuuuuuuuuge :+1: here

**Comment by joshtriplett** [comment]
:+1: here.

**Comment by ibaryshnikov** [comment]
I did some preparations in [this pr](https://github.com/rust-lang/cargo/pull/5611) like filling the boilerplate and updating errors from **cargo-edit** to fit more for **cargo**, but farther work requires some patches to toml parser which is used in cargo (in case we don't want to have two parsers), and currently I don't have enough time to do required updates in the parser on my own. If you find some of the code in this pr fits for the solution, feel free to use it.

**Comment by thejpster** [comment]
I just wanted to add that I think cargo-add is a great tool for people new to Rust, as is cargo-generate. Unfortunately if I recommend them in a guide or tutorial, I have to explain how to install them and that they need to first install GCC *and* cmake *and* libssl-dev, which is non-trivial.

**Comment by withoutboats** [comment]
Definitely would love to see this added to cargo, since my workflow setting up a new dev environment is:

1. Install Rust
2. Install ripgrep
3. `cargo install cargo-edit`

As @ibaryshnikov points out though, its nontrivial work since the TOML parser cargo uses right now doesn't preserve layout, whitespace, or comments. We need to migrate to using the same parser cargo-edit uses before we can upstream any of its features into cargo.

**Comment by withoutboats** [comment]
We discussed this in the cargo meeting. There was strong support for upstreaming the functionality of `cargo add` into `cargo`. We doubt it can be stable in time for the 2018 edition (since the final cargo master for the edition needs to be merged roughly 5 weeks from today) but we do think its good to start making progress on adding this functionality.

We are comfortable having two TOML parsers in cargo for now (with a view to eventually merging them) so long as they share a common lexer, to be confident they aren't interpreting the TOML significantly differently. @alexcrichton says he's left a comment on the repo for the parser cargo-edit is using to this effect.

**Comment by sandangel** [comment]
Hi everyone, may I know the status of this issue? It's 2019 now and Happy New Year. ^^

**Comment by Eh2406** [comment]
From the discord Today at 5:48 PM:
"ehuss: (btw, my format-preserving toml parser just passed all tests :tada: :smile: )"

**Comment by cup** [comment]
https://github.com/rust-lang/cargo/issues/1917
https://github.com/rust-lang/cargo/issues/2179
https://github.com/rust-lang/cargo/issues/4

https://github.com/rust-lang/cargo/issues/5586
https://github.com/rust-lang/cargo/issues/6457
https://github.com/rust-lang/cargo/issues/6488
https://github.com/rust-lang/cargo/issues/651
https://github.com/rust-lang/cargo/pull/5611

**Comment by yhaskell** [comment]
Hello! Does anyone know what is the status of this?

**Comment by cup** [comment]
https://github.com/killercup/cargo-edit/issues/252

**Comment by ordian** [comment]
In https://github.com/rust-lang/cargo/pull/5611#issuecomment-401895847 a concern was raised that we should (partially) deduplicate toml parsing libraries (cargo currently uses toml-rs, but that's not a lossless toml parser, e.g. it will eat your comments, cargo-edit currently uses [`toml_edit`](https://github.com/ordian/toml_edit)) before using it in cargo. Unfortunately, I don't have time to work on it [(link to issues)](https://github.com/ordian/toml_edit/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.2), but would be happy to mentor if someone wants to make this happen (if you're interested, please leave a message in https://gitter.im/toml_edit/Lobby).

After the above issues are resolved, we can hopefully revive #5611.

**Comment by thomasharmon808** [comment]
@SamedhG and I have started to work on updating [`toml_edit`](https://github.com/ordian/toml_edit) to support [`toml v0.5`](https://github.com/toml-lang/toml/blob/master/versions/en/toml-v0.5.0.md) We wanted to hear the opinions of the rust team and see which is a better option, or if there is not a significant difference between the two:

1. Add lossless parsing to [`toml_rs`](https://github.com/alexcrichton/toml-rs) and adapt [`cargo_edit`](https://github.com/killercup/cargo-edit) to use that instead of `toml_edit`
2. Update `toml_edit` to support `toml v0.5` and update `cargo` to use `toml_edit`

**Comment by ehuss** [comment]
I'd be very wary of having a second TOML parser to Cargo.  I think it was previously mentioned that if they could share a lexer, that might be a compromise.  I'm still concerned about adding a large dependency that has doubt around how well it is maintained.

I did work on an experimental preserving API for toml-rs, but never finished it.  I do have the intention of picking it back up, but probably not for a long while.

Creating a good, clean API that handles all the intricacies of toml is quite difficult, and it will require some moderately invasive changes to toml-rs.  I unfortunately don't have a lot of time to help with this right now.

**Comment by ordian** [comment]
I agree with @ehuss there. On the other hand, I don't know how different the first approach would be in terms of maintenance, because in the end it would still be two different toml libraries sharing the same lexer and maybe some other parts. Even though I don't have time to work on some features myself, I still can review PRs and accept contributions.

**Comment by ehuss** [comment]
I have pushed an updated branch of the changes I was working on toml-rs for providing a raw API ([raw-api](https://github.com/ehuss/toml-rs/tree/raw-api)). The idea is to have a low-level API exposed in the toml deserializer, and then a format-preserving API can be built on top of it.  If someone wants to help push this along, one possibility is to build that API on top of that raw API and see how it goes from there.  I had started on a document API earlier this year ([document](https://github.com/ehuss/toml-rs/tree/document)) that uses the raw API, but I think it ended up being too ambitious, trying to automatically handle all the table kinds in a an intelligent way.  

@alexcrichton What do you think about that strategy, and the general shape of the API changes to toml-rs?  In essence, [`Deserializer::fragment`](https://github.com/ehuss/toml-rs/blob/c9d1b07c069ce32fc5630bdaabca2381be3c8431/src/de.rs#L1372-L1387) is the same as the old `line` function, with `pub` added.  Each [`Fragment`](https://github.com/ehuss/toml-rs/blob/c9d1b07c069ce32fc5630bdaabca2381be3c8431/src/de.rs#L2377-L2393) kind retains all the whitespace/comments of the surrounding text.

**Comment by alexcrichton** [comment]
Looks great to me! I'd be down for adding that and trying to build `cargo-edit` on top of that.

**Comment by ordian** [comment]
Just to give an update from my side: I didn't have enough spare to work on toml_edit myself (but hopefully this will change soon), but there is an open PR to add dotted keys support. The thing is it's tricky to make an easy to use and lossless API to work with dotted keys, at least it requires some thought and time.

I would be happy to unify and merge our efforts @ehuss, so that we can make this happen sooner. Also, there is a need for lossless toml parser outside of cargo-edit as well, e.g. for cargo-release, cargo-tidy, cargo-audit-fix, etc. It would be a shame if it will end up being a one-man job. Let's collaborate to increase the bus factor!

**Comment by tyt2y3** [comment]
@ehuss Can you name a few intricacies of TOML you think is difficult to handle? I'd like to assess how difficult the task would be, if I or someone would be able to help.

**Comment by ehuss** [comment]
Some of the tricky bits I can think of:

* 4 different table types (bracketed tables, inline tables, dotted keys, and array-of-tables).  Generally when editing, you want to preserve the existing style, and insert into the correct location.  I wanted to make this transparent so that the user doesn't need to do a lot of work to detect which is the correct type to use, and where it can be placed.  
* Array-of-tables are contextual, with tables defined after an array belonging to the most recent table.
* It would be nice to have an API that supported index operations like serde_json's `Value` type does, but since TOML doesn't have a null/none concept, that is quite tricky. It looks like `toml_edit` does this, too.

I would suggest starting with the raw API I created, and creating something very simple that can convert the raw elements into a structured document with a read-only API, and a way to re-render the document.  Then try to add simple mutation methods that don't break the structure, and incrementally try to create more sophisticated APIs that are easier to use and more natural.  I would probably look at `toml_edit` and at least use it as inspiration if not make it identical or modify it to use the `toml` crate.

**Comment by DrYSG** [comment]
How is this ticket different than: https://crates.io/crates/cargo-edit

**Comment by steveklabnik** [comment]
This ticket is about up streaming that, at least conceptually.

**Comment by kornelski** [comment]
It looks like this highly-desirable feature has got delayed for years waiting for a super-perfect format-preserving TOML reparser.

IMHO perfection in the parser is not that important. I've been using `cargo-edit` even when it had non-preserving parser and sorted all the keys alphabetically.

Non-preserving parser is almost a feature! It's like `cargo fmt` for TOML files :)

**Comment by lig** [comment]
@kornelski Personally, I'd prefer the one which sorts everything as I do it myself every time after cargo-add:)

**Comment by joshtriplett** [comment]
The current version of cargo add will keep your dependencies in sorted order as long as they started out in sorted order.

**Comment by Boscop** [comment]
@joshtriplett It doesn't seem to be the case (?)
I had this:
```toml
[dependencies]
persistence = { path = "../persistence" }
```
then I did `cargo add serde serde_derive` and then `cargo add anyhow`. Then it looked like this:
```toml
[dependencies]
persistence = { path = "../persistence" }
serde = "1.0.118"
serde_derive = "1.0.118"
anyhow = "1.0.37"
```
Why didn't it add `anyhow` before the other deps?

**Comment by JerryGreen** [comment]
This command should be built-in, that's for sure. For the rest people, there's a temporary solution:

```bash
cargo install cargo-edit # installs "cargo add" command
cargo add image # adds "image" package into dependencies
```

**Comment by joseph-henry** [comment]
Any recent movement on this?

**Comment by epage** [comment]
I'm willing to pick up this work and see it through.

I have thoughts on the conversation on `toml` vs `toml_edit` but first wanted to check in on one approach I've not seen mentioned: what if we implement `serde::Deserializer` for `toml_edit`, get it toml 1.0 compliant, and switch over?

**Comment by joshtriplett** [comment]
@epage That seems like a reasonable approach to me.

**Comment by ordian** [comment]
I think the initial plan was to
1. Make the toml's [lexer](https://github.com/alexcrichton/toml-rs/blob/master/src/tokens.rs) public and reuse in toml-edit.
2. Get `toml_edit` 1.0 compliant.
3. Switch over to `toml_edit` where it makes sense.

I don't know how well `serde` model plays with format-preserving (incl. indentation) parsers, but if it's doable and the only reason stopping the complete switch-over, I don't see why not.

P.S. I'll be offline for the 2 weeks starting 6th of Aug, but happy to accept PRs when I'm back.

**Comment by epage** [comment]
> I think the initial plan was to
> Make the toml's lexer public and reuse in toml-edit.

I figured this was a faster path to 1.0.

However, if the best way to getting `toml_edit` to being 1.0 compliant is to switch the lexer, I am game for looking into it.  I don't want to do throwaway work.

If its not relevant for 1.0, I'm curious how much is gained for the two crates and end-users from sharing a lexer?  Are there any costs and how bad are they?   From my outside perspective, it seems like only little is being shared but we are now coupling architectures.

> I don't know how well serde model plays with format-preserving (incl. indentation) parsers, but if it's doable and the only reason stopping the complete switch-over, I don't see why not.

If we are just deserializing, then format preserving shouldn't impact things.  I do have some more out-there ideas for serializing by doing tree-diffing so we only modify the parts that are changed, leaving the rest to be the same.  It'd be a great ergonomic improvement in cases like `cargo-release` where all we care about is updating a version number.  We just can't hand-crate the format for other cases, like tables.

> P.S. I'll be offline for the 2 weeks starting 6th of Aug, but happy to accept PRs when I'm back.

Enjoy your vacation!

**Comment by ehuss** [comment]
> what if we implement serde::Deserializer for toml_edit

Can you say more why you would want to use serde?  I suspect trying to shuttle information through serde's data model would make things a bit more difficult, and I'm not sure how that relates to supporting cargo-edit.

> Nothing stops us from switching _back_ to toml-rs if it grows a format-preserving API in the future. But in the meantime, switching to toml_edit seems like it would address the concerns about duplicate toml parsers.

I'm not sure I understand this.  There is an outline given [above](https://github.com/rust-lang/cargo/issues/5586#issuecomment-562978274) on how to proceed, and it involves updating toml-rs to provide a raw API upon which something like toml_edit could depend on.  There's a branch linked there with most of the work done, and then there is the other half of of work to consume that raw output and provide ways to mutate it.

**Comment by epage** [comment]
> Can you say more why you would want to use serde? I suspect trying to shuttle information through serde's data model would make things a bit more difficult, and I'm not sure how that relates to supporting cargo-edit.

It looks like Cargo is using serde with toml-rs today.  If toml-edit supported serde, it would be a drop-in replacement.  cargo-edit would continue to use the lower level, format-preserving, API.  We then meet the goal of having one toml parser for both needs.

I fully admit I am probably glossing over cases that I haven't come across yet but I assume we can make them work in this model.

**Comment by ehuss** [comment]
I think for now it would be desirable to keep Cargo's existing code separate from cargo-edit functions.  That is, nothing in Cargo's existing manifest parsing should need to change.  My understanding of cargo-edit is that it has pretty basic needs of finding specific tables (like "dependencies"), and it doesn't need to do the same things that the rest of Cargo does in terms of consuming the TOML files.

**Comment by ssokolow** [comment]
@ehuss For clarity, are you arguing that, for now, it's better to prioritize stability to the point of linking two independent TOML parsers into Cargo, rather than retrofitting the implementation of the current API Cargo uses so a format-preserving lower-level API can be tapped off the side of it for cargo-edit to use?

**Comment by epage** [comment]
> I think for now it would be desirable to keep Cargo's existing code separate from cargo-edit functions. That is, nothing in Cargo's existing manifest parsing should need to change.

In case there is a misunderstanding, I want to be clear.  The only changes to existing cargo code I am proposing are

- We switch out the toml `serde::Serializer` from `toml` to `toml-edit`
- If a part of cargo doesn't use serde for toml, we switch to using serde

The `cargo-edit` code would not try to use any of the existing serde-based manifest logic inside of `cargo` but would keep its existing use of the lower level API.

While I did mention an idea for making it possible to use serde to make format-preserving edits to toml, that was only intended as an unrelated aside and was meant more for discussing toml-edit's API and not saying cargo should use it (it'd work well for `cargo-release` but `cargo` might have more problems with it).  Experimenting with that is not within the scope of what I propose we do.

**Comment by ehuss** [comment]
>  prioritize stability to the point of linking two independent TOML parsers into Cargo,

Not exactly.  The idea is to share the same low-level parser between toml-rs (the raw API suggested above) and toml-edit. There should be essentially one parser.  It's the high-level API that would be different (the existing serde API in Cargo, and the toml-edit API for cargo-edit).

**Comment by epage** [comment]
> There should be essentially one parser

This was the stated goal, though I think there is an unstated goal lying behind it.  No idea if it is faster compile times, ensuring consistent parse behavior, consistent error reporting, or something else.  Knowing that would be a big help.

Exclusively using toml-edit meets the stated goal.

Sharing lexers is another way to the stated goal.  I fully admit my lack of familiarity, but it seems like it would only partially meet that goal because a lot would still be left to the individual crates on top.

**Comment by ehuss** [comment]
The main goals is to have consistent parsing behavior, and to minimize the surface of what needs to be maintained.  The amount of work the upper-layer needs to do should be minimal.  The hardest part is probably merging dotted keys into internal maps.  Since the way serde works is a lot different than what an editing API wants, I don't think that can be easily merged, but those are the kinds of questions that the implementer needs to explore.

We also have a desire to use the lower-level parsing APIs for parsing partial TOML values in other places of Cargo, such as the `cargo config` command.  A low-level parsing API would be beneficial for that, too.

**Comment by epage** [comment]
>  The amount of work the upper-layer needs to do should be minimal. The hardest part is probably merging dotted keys into internal maps

Could you help me understand how sharing a common base would help towards this?  So far, the talk I've seen is for merging lexers.  I feel like I'm missing something because I would expect a lexer to offer minimal help towards this goal.  I could understand more if the push was to share a format-preserving push-parser, then I could see both libraries being able to build on top of that.  I've not dealt with the parsing intricacies of toml yet, so I'm unsure if there are still areas this wouldn't be able to help with, like merging dotted keys.

Now, I could see building toml-rs directly on toml-edit since all it needs to do so throwaway data to convert from toml-edit to toml-rs.

**Comment by kornelski** [comment]
Can you separate addition of `cargo add` from addition of TOML format-preserving, lexer merges, parser deduplication, etc.?

There's a huge user need for *any* support for *basic* functionality, even if it's format-destroying TOML-rewriting tool. It's an optional command that users don't have to use, so if anyone had a complex TOML file that couldn't be rewritten, they can just continue updating their TOML manually. But for everyone else going through a Rust basics tutorial or a README of a dependency, having `cargo add` in any form would already be a big improvement.

**Comment by joshtriplett** [comment]
I don't think we should break that functionality of cargo-add when adding it. There are at least two proposals in this thread of ways we can add cargo-add and keep that functionality.

**Comment by PoignardAzur** [comment]
@epage Any progress?

I'm aware any PR solving this will probably be blocked because of non-trivial disagreements on which toml dependency should be used; but in the meantime, is there a fork somewhere with a proof of concept?

Either way, is there additional work I could help with?

**Comment by epage** [comment]
> I'm aware any PR solving this will probably be blocked because of non-trivial disagreements on which toml dependency should be used;

We've [discussed this on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/toml.2C.20toml_edit.2C.20and.20cargo-add).  The concerns are
- Validation work
- Performance: at worst only having a negligible performance impact when parsing a large `Cargo.toml` dependency tree

I'm moving forward with `toml_edit` assuming it will be able to satisfy both cases

> Any progress?

So far I've
- Made the [toml validation suite](https://github.com/BurntSushi/toml-test) a [rust test harness](https://github.com/epage/toml-test-rs)
- [Posted a PR to add that harness to toml_edit](https://github.com/ordian/toml_edit/pull/111)
- Have several toml 1.0 conformance branches stacked on top of that ready to post as PRs
- Have an experimental local branch for `serde:Deserialize`

Progress is limited by me splitting my time also with getting clap3 out the door and the toml_edit maintainer being on a couple week vacation.

> Either way, is there additional work I could help with?

My expectations for future work include
- Finish toml 1.0 conformance work
  - I think its mostly date/time and dotted keys (started small to get familiar with the crate and its parser)
- Effectively porting all of `toml-rs` on top of `toml_edit`
- Ensuring `toml_edit` (through a feature flag?) can parse any non-conformant `Cargo.toml` files on crates.io
  - Probably require downloading all crates and comparing the result of each parser on every Cargo.toml file using `toml_test::encoded::Encoded`

Since I've gotten started on the parser, it'd be easier for me to finish that up.  So a good place to help out would be:
1. Make a `toml_test::encoded::Encoded` conversion for `toml-rs`, possibly by contributing the 1.0 conformance test harness to `toml-rs` (ignoring existing failures)
  - Will be needed to verify the `toml-rs` compatible API on top of `toml_edit`
  - Will be needed if confirm non-conformance support by comparing the result of each crate across all of crates.io
  - Note: date/time support might need some work, the upstream conformance suite does some special handling with those
2. Pick up my serde work and finish porting all of `toml-rs` on top of `toml_edit` (I can share my serde  branch

**Comment by epage** [comment]
So I have a more accurate idea of what work remains, **what are the expectations for adding a cargo sub-command to the Rust toolchain?**

`cargo-add` does not depend directly on `cargo` and does almost everything its own way
- Arg parsing with `structopt`
- Custom user messages, without `--message-format` support
- Relies on `cargo_metadata` and `crates-index`
- Makes network connections through `ureq` with the help of `env_proxy`
- Uses `toml_edit` (already discussed, just including for completeness)
- Tests are written using `assert_cmd` and `assert_fs`

The only thing I see in common is high level user behavior and the use of `termcolor`

I've also been polishing up the behavior of `cargo-edit` but wanted to call out [the issue list](https://github.com/killercup/cargo-edit/issues?q=is%3Aissue+is%3Aopen+label%3Acargo-add) in case there are any that they cargo team considers a priority before merging.

**Also, is there a recommendation for how to review and merge a command?**  I'm assuming a code drop of the entire command as a single PR would not be ideal.  If there is a way to mark an entire command as nightly-only, we could start with a top-level skeleton with everything else stubbed out and PR by PR fill out those stubs.  Its also extra work to do something like this, so I'd be interested in recommendations in how to balance between submitter and reviewer.

---

For anyone else following along, this does not necessarily mean we are "almost there

Whats done:
- `toml_edit` supports almost all of TOML 1.0
- `toml_edit` is nearly a drop-in replacement for `toml-rs`
- `cargo` has dropped support  for historical toml incompatibilities, so `toml_edit` won't  need to support them
- `cargo-add` treats paths relative to CWD
- `cargo-add` has improved defaults when adding a workspace member (https://github.com/killercup/cargo-edit/pull/504)

Remaining work
- Further characterization of `toml_edit`s performance and close the gap as needed (https://github.com/ordian/toml_edit/issues/132)
- Find and support any other toml incompatibilities when switching to `toml_edit` (https://github.com/ordian/toml_edit/issues/130)

**Comment by ehuss** [comment]
In terms of expectations for what new commands should follow, here are some rough guidelines:

* Any new dependencies need to have licenses we can use.
* It should roughly follow the same form as other commands. 
* I would probably not want to bring in structopt, so that it can integrate with Cargo's existing argument parsing. 
* I would not expect it to shell out to `cargo metadata` or anything like that, I would expect it to use Cargo's internal APIs.
* Given that it should use Cargo's internal APIs, I would not expect it to need to bring in any networking libraries. 
* We can maybe have some flexibility bringing in additional dependencies for tests, though if there are existing things in cargo's testsuite that can replace it, that would be nice (sharing can also be deferred to later).
* When it is stabilized, it will need to update all the documentation.
* I didn't look at the issue list closely, but I doubt there is anything there that would be a prerequisite.

In terms of starting the merge process, that's a bit tougher to give guidance.  It would probably be good before getting started to make an outline of what is being proposed to merge. https://github.com/rust-lang/cargo/pull/8062#issuecomment-606910003 is an example of a command brought into cargo, and the merge proposal (and the top comment outlined what was changed). 

Commands can be made unstable (see [`cargo config`](https://github.com/rust-lang/cargo/blob/0121d66aa2ef5ffa9735f86c2b56f5fdc5a837a6/src/bin/cargo/commands/config.rs#L30-L32) as an example).  However, since the cargo-edit commands are so widely used, we would need to be careful not to break people's environments.  I'm not sure how to go about that.  Some options:

* Just go straight to stabilized, like we did with cargo-tree.
* Use temporary alternate names, like `cargo unstable-add`.
* Change the priority so that if cargo-edit is installed, it uses that instead of the built-in command (until it is stabilized).

I kinda lean towards the first option, but I guess that depends on the difficulty of getting it all together in one PR.

**Comment by alexcrichton** [comment]
What I would add over what @ehuss already mentioned is that I think the goal is to make a new subcommand feel idiomatic within Cargo's own codebase. If the style is completely different or if the dependencies are completely different then that makes it much more difficult to maintain over time or it just won't receive maintenance as we won't know how to update it. IIRC `cargo tree` was in some ways almost entirely rewritten as it moved into Cargo, so I don't think that adding new subcommands is going to be as easy as copy/paste. I think it's fine to defer work to later but only if there's a clear path of what needs to be done and someone is on the hook for doing it, "this should get fixed by someone in the future" in my opinion would block merging.

**Comment by epage** [comment]
Thanks!

I suspect I'll need to fork `cargo-edit` to pull out `cargo-add` specifically, and work to adapt it to `cargo`s implementation style.  The tests are pretty good and I should be able to do this with minimal or no test change.

**Comment by epage** [comment]
In the past, people have asked what they can do to help.  I have identified some [issues in toml_edit](https://github.com/ordian/toml_edit/labels/cargo) that are regressions compared to toml-rs.  These were found by cargo's tests after porting it over.  

Feel free to comment on ones you are interested and let me know if there is any way I can help in mentoring on these!

**Comment by romain20100** [comment]
Hello there,

Apparently this feature has been awaited for quite some time. No doubt the toml writing part has played a role in that.

I was wondering if the idea of just "printing the line that would be added to Cargo.toml" has been considered.

Finding the latest version of a given crate without leaving the terminal, and have it pre-written in the relevant format should be a good first step that solves most of the developer's trouble.

And if I'm not mistaken, it seems waaay simpler than trying to handle all the possible shenanigans of how an existing file could have been written.

Or did I miss something?

**Comment by epage** [comment]
I think going that route depends on the answer to the questions like:
- Can we change behavior away from that once toml_edit is ready?  My guess is "no".
- Will taking over the "cargo add" name with diminished functionality be acceptable by end users?  My guess is "no"

With that said, I was considering the idea of creating a command for interacting with the registry that could possibly fill the role of what you mentioned.  Right now, cargo has its own copy of a lot of logic but isn't very good for depending on for external tools.  Some of the logic gets reused through `cargo metadata` command which is wrapped by the `cargo_metadata` crate.  A lot of other functionality gets duplicated instead, like `crates_index` crate.  What if we instead made a cargo command for updating the local index and querying it and wrapping that in a crate?  My hope is this would be an acceptable way for cargo to expose the registry code for reuse.  You could then use this command as well to look up crates.

Also, https://github.com/rust-lang/cargo/pull/10086 now passes tests.  All that is left is to audit the user facing changes (and adjust as needed) and I need to do benchmarking.

**Comment by Alexendoo** [comment]
@romain20100 Hi!

You can use `cargo search` for this, the results it returns are valid to be put in a Cargo.toml

**Comment by romain20100** [comment]
@Alexendoo Sure, it's relatively trivial to adapt the result. In fact I've done just that with a simple bash function (see below).
But it's not immediately usable.

My point was just that there may be an intermediate milestone that could be simpler to achieve.
Although it's possible we are now close to a full on solution. In that case it's ok.

---

```bash
function cargo-add {
    output=$(cargo search $1)
    tmp="${output//$1 = \"}"
    version="${tmp//\"*}"
    echo "$1 = { version = \"$version\" }"
}
```

**Comment by lnicola** [comment]
@romain20100 you can already copy-paste the output of `cargo search` to `Cargo.toml`. E.g. `serde = "1.0"` is enough, you don't have to write it as `serde = { version = "1.0" }`.

I don't think there's any need for an intermediate step here. `cargo edit` can and will be added to `cargo` proper when it's considered ready.

**Comment by simonsan** [comment]
> serde = { version = "1.0" }

At least from `cargo-smart-release` I know from the [limitations section](https://github.com/Byron/gitoxide/tree/main/cargo-smart-release#limitations), that the version needs to be inside a table though.

**Comment by epage** [comment]
> > serde = { version = "1.0" }
> 
> At least from `cargo-smart-release` I know from the [limitations section](https://github.com/Byron/gitoxide/tree/main/cargo-smart-release#limitations), that the version needs to be inside a table though.
> 
> Edit: By which I mean that there is sometimes this requirement, which probably should be put behind an option flag or somethign? thinking Not sure.

imo the focus should be on fixing that git-smart-release limitation.  The code exists for it in [`cargo-release`](https://github.com/sunng87/cargo-release) and in `cargo-set-version` (in this repo).

**Comment by Byron** [comment]
Thanks for bringing up the issue @simonsan , I totally agree with @epage and think smart-release should learn to deal with non-table versions.

**Comment by epage** [comment]
As an update, https://github.com/rust-lang/cargo/pull/10086 has been merged which was the major blocker for cargo-edit being upstreamed.  

I am now preparing cargo-add for merging.
- I've upgraded cargo-edit to the latest toml_edit, past some breaking changes that had a large effect on it
- I am currently porting the tests to `trycmd` which, through its snapshot updating, will make it easier to update tests as we port cargo-add to use cargo's facilities
- Next, I will fork cargo-add into a branch and will slowly refactor it into a state for merging
  - Isolating all of the logic `cargo-add` needs from the `cargo-edit` `[lib]`
  - Port parts of `cargo-add` from what was in  `cargo-edit` `[lib]`  to `cargo`
- Then we'll be ready for a draft PR with the proposal to merge, like [cargo tree](https://github.com/rust-lang/cargo/pull/8062#issuecomment-606910003)
  - I think an important part of this will be calling out the current CLI and re-examining parts of it  (I think `--upgrade` in particular should be re-examined for how to expose that functionality)
  - I feel like this issue has too much noise in it already to have the conversation.  Maybe it could be had on internals?
- We can then start prioritizing and preparing the next `cargo-edit` command

**Comment by epage** [comment]
FYI I just posted on [internals](https://internals.rust-lang.org/t/feedback-on-cargo-add-before-its-merged/16024) looking for feedback before we restrict cargo-add's evolution by merging it in cargo.

**Comment by nyurik** [comment]
Is porting `cargo upgrade` part of this ticket, or would that be a separate one? I couldn't find any mentions of it besides this issue. Thx!

**Comment by epage** [comment]
The plan is to focus on `cargo add` as its seen as the highest priority.  After that, I'll be looking to `cargo rm` because of the symmetry with `add` and it being a small lift.  `cargo upgrade` will be after that and there are a lot of things we need to figure out (e.g. should it be kept as a separate command or expressed as flags on `cargo update`)

**Comment by epage** [comment]
Once I merge this last PR, the `merge-add` branch is complete.  Tomorrow, I expect to try to move the code over to cargo and make any adjustments needed for being in-tree.  After that, I'll be making the PR that proposes merging cargo-add into cargo.

**Comment by epage** [comment]
[Its live!](https://github.com/rust-lang/cargo/pull/10472)
